src/magic_folder/magic_folder.py

```diff
@@ -838,3 +852,141 @@ def add_file(self, path):
         d = defer.Deferred()
         self._queue.put((bytespath, d))
         return d
+
+
+class IRemoteSnapshotCreator(Interface):
+    """
+    An object that can create remote snapshots representing the same
+    information as some local snapshots that already exist.
+    """
+    def upload_local_snapshots():
+        """
+        Find local uncommitted snapshots and commit them to the grid.
+
+        :return Deferred[None]: A Deferred that fires when an effort has been
+            made to commit at least some local uncommitted snapshots.
+        """
+
+
+@implementer(IRemoteSnapshotCreator)
+@attr.s
+class RemoteSnapshotCreator(object):
+    _state_db = attr.ib()
+    _local_author = attr.ib()
+    _tahoe_client = attr.ib()
+    _upload_dircap = attr.ib()
+
+    @eliotutil.inline_callbacks
+    def upload_local_snapshots(self):
+        """
+        Check the db for uncommitted LocalSnapshots, deserialize them from the on-disk
+        format to LocalSnapshot objects and commit them into the grid.
+        """
+
+        # get the mangled paths for the LocalSnapshot objects in the db
+        localsnapshot_names = self._state_db.get_all_localsnapshot_paths()
+
+        # XXX: processing this table should be atomic. i.e. While the upload is
+        # in progress, a new snapshot can be created on a file we already uploaded
+        # but not removed from the db and if it gets removed from the table later,
+        # the new snapshot gets lost. Perhaps this can be solved by storing each
+        # LocalSnapshot in its own row than storing everything in a blob?
+        # https://github.com/LeastAuthority/magic-folder/issues/197
+        for name in localsnapshot_names:
+            action = UPLOADER_SERVICE_UPLOAD_LOCAL_SNAPSHOTS(relpath=name)
+            try:
+                with action:
+                    yield self._upload_one_snapshot(name)
+            except:
```

Comment:
Yes on `except Exception:` I guess.  I'm never sure what the practical difference is.  Is it _good_ to let `KeyboardInterrupt` pass through this layer?  Dunno ...

There is logging already, at least.  The exception has to propagate through the `action` context manager and it will log the failure nicely (marking the action as failed and including a traceback).  So yea, extra magic. :)